model/stats/db.go

```diff
@@ -442,15 +443,350 @@ func statsToUpdatePipeline(projectId string, start time.Time, end time.Time) []b
+		"then": boundaries[lastIndex],
+	}
+	return bson.M{"$addFields": bson.M{fieldName: bson.M{"$switch": bson.M{"branches": branches}}}}
+}
+
+// Builds the _id field for the $group stage corresponding to the GroupBy value.
+func buildGroupId(groupBy GroupBy) bson.M {
+	id := bson.M{"date": "$date"}
+	switch groupBy {
+	case GroupByDistro:
+		id["distro"] = "$_id.distro"
+		fallthrough
+	case GroupByVariant:
+		id["variant"] = "$_id.variant"
+		fallthrough
+	case GroupByTask:
+		id["task_name"] = "$_id.task_name"
+		fallthrough
+	case GroupByTest:
+		id["test_file"] = "$_id.test_file"
+	}
+	return id
+}
+
+// Edits a match expression to include that the 'fieldName' field must have its value in 'values'.
+// Does nothing if 'values' is empty.
+//func addMatchIn(matchExpr bson.M, fieldName string, values []string) {
+//	if len(values) == 1 {
+//		matchExpr[fieldName] = values[0]
+//	} else if len(values) > 1 {
+//		matchExpr[fieldName] = bson.M{"$in": values}
+//	}
+//}
+
+func buildMatchArrayExpression(values []string) interface{} {
+	if len(values) == 1 {
+		return values[0]
+	} else if len(values) > 1 {
+		return bson.M{"$in": values}
+	}
+	return nil
+}
+
+// Adds to an existing $match expression the conditions imposed by the filter StartAt field
+func buildTestPaginationOrBranches(filter *StatsFilter) []bson.M {
+	var dateOperator string
+	if filter.Sort == SortLatestFirst {
+		dateOperator = "$lt"
+	} else {
+		dateOperator = "$gt"
+	}
+
+	var fields []paginationField
+
+	switch filter.GroupBy {
+	case GroupByTest:
+		fields = []paginationField{
+			{Field: "_id.date", Operator: dateOperator, Value: filter.StartAt.Date},
+			{Field: "_id.test_file", Operator: "$gt", Value: filter.StartAt.Test},
+		}
+	case GroupByTask:
+		fields = []paginationField{
+			{Field: "_id.date", Operator: dateOperator, Value: filter.StartAt.Date},
+			{Field: "_id.task_name", Operator: "$gt", Value: filter.StartAt.Task},
+			{Field: "_id.test_file", Operator: "$gt", Value: filter.StartAt.Test},
+		}
+	case GroupByVariant:
+		fields = []paginationField{
+			{Field: "_id.date", Operator: dateOperator, Value: filter.StartAt.Date},
+			{Field: "_id.variant", Operator: "$gt", Value: filter.StartAt.BuildVariant},
+			{Field: "_id.task_name", Operator: "$gt", Value: filter.StartAt.Task},
+			{Field: "_id.test_file", Operator: "$gt", Value: filter.StartAt.Test},
+		}
+	case GroupByDistro:
+		fields = []paginationField{
+			{Field: "_id.date", Operator: dateOperator, Value: filter.StartAt.Date},
+			{Field: "_id.variant", Operator: "$gt", Value: filter.StartAt.BuildVariant},
+			{Field: "_id.task_name", Operator: "$gt", Value: filter.StartAt.Task},
+			{Field: "_id.test_file", Operator: "$gt", Value: filter.StartAt.Test},
+			{Field: "_id.distro", Operator: "$gt", Value: filter.StartAt.Distro},
+		}
+	}
+
+	return buildPaginationOrBranches(fields)
+}
+
+// Creates an aggregation pipeline to query task statistics.
```

Comment:
Nit: Comment style for this function and below.

Reply:
Done.